lib/tic_tac_toe.ex

```diff
@@ -2,16 +2,22 @@ defmodule TicTacToe do
   alias TicTacToe.Display
   alias TicTacToe.Player
   alias TicTacToe.BoardUpdate
+  alias TicTacToe.BoardInspect
 
-  def run(%{board: %{game_over: game_over}} = state) when game_over do
-    show_board(state)
+  def run(state) do
+    if game_over?(state) do
+      show_board(state)
+    else
+      state
+      |> show_board()
+      |> play_turn()
+      |> run()
```

Comment:
You probably already did this research but it turns out Elixir does have tail call optimization here, if you're worried about it you don't have to be.

https://elixir-lang.org/getting-started/recursion.html

Reply:
Thanks